lib/dom/xml_entities.nit

```diff
@@ -0,0 +1,586 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# This file is free software, which comes along with NIT.  This software is
+# distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+# without  even  the implied warranty of  MERCHANTABILITY or  FITNESS FOR A
+# PARTICULAR PURPOSE.  You can modify it is you want,  provided this header
+# is kept unaltered, and a notification of the changes is added.
+# You  are  allowed  to  redistribute it and sell it, alone or is a part of
+# another product.
+
+# Basic blocks for XML parsing
+#
+# DOM entities are defined in this module, specifically:
+#
+# * XMLEntity: Abstract kind of XML-related node
+# * XMLDocument: A well-formed XML document, root of the tree
+# * PCDATA: Raw XML-escaped character data
+# * CDATA: Raw data, may contain invalid XML escape characters
+# * XMLTag: Abstract XML tag element
+# * XMLAttrTag: Abstract XML element, they may contain attributes
+# * XMLOnelinerTag: Any tag contained on one-line only
+# * XMLStartTag: A tag starting a new hierarchy level in the tree
+# * XMLPrologTag: A tag containing meta-information on the document
+# * XMLEndTag: A tag signaling the end of a block
+# * XMLCommentTag: A comment tag
+# * XMLSpecialTag: A special tag, which may contain meta-information (DOCTYPE for instance)
+# * XMLAttribute: Any kind of attribute that may be attached to a tag
+# * XMLStringAttr: An attribute containing a String
+# * XMLError: Any kind of error thrown while parsing a document
+module xml_entities
+
+# Information about the location of an entity in an XML source document
+private class Location
+	# Line in which the element is described
+	var line: Int
+	# Offset in the line at which the element is positioned
+	var offset: Int
+
+	redef fun to_s do return "line {line}, position {offset}"
+end
+
+# Any kind of XML Entity
+abstract class XMLEntity
+	# Optional parent of `self`
+	var parent: nullable XMLEntity is private writable(set_parent)
+
+	# Optional location of the entity in source
+	private var location: nullable Location
+
+	# The children of `self`
+	var children = new HashSet[XMLEntity]
+
+	# Sets the parent of `self` to `e`
+	fun parent=(e: XMLEntity) do
+		if parent != null then
+			parent.children.remove(self)
+		end
+		set_parent(e)
+		e.children.add(self)
+	end
+end
+
+# Top XML Document-Object Model element, contains all the parsing facilities
+class XMLDocument
+	super XMLEntity
+
+	redef fun to_s do
+		var s = ""
+		for i in children do s += i.to_s
+		return s
```

Comment:
Truth is, it depends on the size of the elements, as the benchmark of array_to_s proved, although, as a general rule, when elements are small or of medium size, the malloc strategy proved to be the fastest.

But in several cases where the output is large, like this, I think a Rope might be faster, but as there is no warranty of this, I guess it can indeed be replaced by a `join`